proposals/0000-no-toplevel-field-selectors.rst

```diff
@@ -0,0 +1,134 @@
+
+There have been proposals to extend the usage of records in Haskell
+(DuplicateRecordFields, OverloadedRecordFields, DeriveGeneric to name a few),
+but none of them address the original issue where record fields steal the
+toplevel function name for each field for selector purposes. This proposal
+intends to open the design space to this issue by removing the generation of
+these toplevel selector functions.
+
+Possible Alternative Designs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+- A `generic-lens` equivalent for toplevel lenses
+- namespaced accessors
+- namespaced lenses
+- overloaded-labels based accessors / lenses
+- row types
+- https://github.com/ghc-proposals/ghc-proposals/pull/158
+- ...
+
+Proposed Change Specification
+-----------------------------
+
+Record definition no longer defines toplevel selector functions when the
+``NoToplevelFieldSelectors`` extension is enabled. A
+``NoToplevelFieldSelectors`` pragma for records will also be available, such
+that this extension can be enabled on a per-record basis.
+
+Record construction/update syntax and pattern matching will work as before, the
+disambiguation handled by ``DuplicateRecordFields``. The necessary selectors
+will be part of a ``Record(..)`` export, or can also be named one by one. They
+always have to be associated with the data type though, because there is no more
+toplevel selector.
+
+A function for Template Haskell to go from a record field name to a selector for
+that field will be provided, because it's not possible anymore to go from field
+name directly to selector.
+
+Example
+^^^^^^^
+
+Given a data structure
+
+    data Foo = Foo { bar :: Int, baz :: String }
+
+The following will be available:
+
+- the type ``Foo``
+- the constructor ``Foo``
+- the two functions ``bar`` and ``baz``
+- the names ``bar`` and ``baz`` for record construction (``Foo { bar = 3, baz = "foo" }``)
+- the names ``bar`` and ``baz`` for ``RecordWildCards``
+
+If the language extension ``NoToplevelFieldSelectors`` is enabled for the module
+or ``Foo`` specifically, all of the above will be generated, except for the two
+functions ``bar`` and ``baz``.
+
+Wildcard exports will work as before, except for the two functions. Even if
+these functions are otherwise defined, the wildcard will not export them.
+Exporting the names for record construction now has to be specific to the
+record. Without ambiguitiy, previously this was equivalent
+
+    module A where (Foo(Foo, bar, baz))
+
+    module A where (Foo(Foo, bar), baz)
+
+Because of the new semantics, these two export statements are now different.
+Maybe a wildcard export will export functions of the same name as well, this
+question isn't resolved yet.
+
+Effect and Interactions
+-----------------------
+
+Breakage estimation
+^^^^^^^^^^^^^^^^^^^
+
+Enabling this extension will break a lot of Template Haskell. Going from record
+field name to selector won't work anymore. A new way to go from record field
+name to selector has to be found.
```

Comment:
Can you give some examples of what you expect to break?

Reply:
Anything relying on using `Name` to resolve a datatype accessor. https://github.com/ghc-proposals/ghc-proposals/pull/160#discussion_r215764916 - e.g. I'd guess the codecs Aeson generates. I I'm trying to understand on how it works, but currently failing. Not 100% sure if it actually poses a problem.